cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+  // Only call `Aws::ShutdownAPI`, if `Aws::InitAPI` was called on construction.
+  const bool _shutdown_s3_api;
+
+ public:
+  /**
+   * @brief Create a context given an existing S3 client
+   *
+   * The S3 SDK isn't initialized.
+   *
+   * @param client The S3 client
+   */
+  S3Context(std::shared_ptr<Aws::S3::S3Client> client)
+    : _client{std::move(client)}, _shutdown_s3_api{false}
+  {
+    if (!_client) { throw std::invalid_argument("S3Context(): S3 client cannot be null"); }
+  }
+
+  /**
+   * @brief Create a new context with a newly created S3 client.
+   *
+   * The S3 SDK is automatically initialized on construction and shutdown on destruction.
+   *
+   * The new S3 client use the default `Aws::Client::ClientConfiguration`, thus please make sure
+   * that AWS credentials have been configure on the system. A common way to do this, is to set the
+   * environment variables: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+   *
+   * Other relevant options are `AWS_DEFAULT_REGION` and `AWS_ENDPOINT_URL`, see
+   * <https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html>.
+   *
+   * @param endpoint_override If not empty, the address of the S3 server. Takes precedences
+   * over the `AWS_ENDPOINT_URL` environment variable.
+   */
+  S3Context(const std::string& endpoint_override = "") : _shutdown_s3_api{true}
+  {
+    // NB: `Aws::InitAPI` has to be called before everything in the SDK beside `Aws::SDKOptions`,
+    // even before config structs like `Aws::Client::ClientConfiguration`.
+    // Notice, we may call `Aws::InitAPI` multiple times, which is allowed see:
+    // <https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Aws.cpp#L32>
+    Aws::SDKOptions options;
+    Aws::InitAPI(options);
+
+    // Create a client config where `endpoint_override` takes precedences over `AWS_ENDPOINT_URL`
+    Aws::Client::ClientConfiguration config;
+    const char* ep = std::getenv("AWS_ENDPOINT_URL");
+    if (!endpoint_override.empty()) {
+      config.endpointOverride = endpoint_override;
+    } else if (ep != nullptr && !std::string(ep).empty()) {
+      config.endpointOverride = ep;
+    }
+
+    // We check authentication here to trigger an early exception.
+    Aws::Auth::DefaultAWSCredentialsProviderChain provider;
+    if (provider.GetAWSCredentials().IsEmpty()) {
+      throw std::runtime_error("failed authentication to S3 server");
+    }
+    _client = std::make_shared<Aws::S3::S3Client>(config);
+  }
+
+  ~S3Context() noexcept
+  {
+    if (_shutdown_s3_api) {
+      try {
+        Aws::SDKOptions options;
+        Aws::ShutdownAPI(options);
+      } catch (const std::exception& e) {
+        std::cerr << "~S3Context(): " << e.what() << std::endl;
+      }
+    }
+  }
+
+  /**
+   * @brief Get a reference to the S3 client.
+   *
+   * @return S3 client.
+   */
+  Aws::S3::S3Client& client() { return *_client; }
+
+  // No copy and move semantic
```

Comment:
OK, so the user is always going to hand around const references to this?

Reply:
yes

Reply:
We could support _move_ as long as we make `_shutdown_s3_api` non-const and sets it to `false` to avoid double `Aws::ShutdownAPI` calls

Reply:
but since `RemoteHandle` takes a shared pointer, I expect users to create the `S3Context` in a shared pointer as well